### DISCUSSION TITLE
What are the reasons why business logic is sometimes placed in state class instead of a service?

### DISCUSSION THREAD

**Discussion by vgpro54321** [ngxs/store#1816] [Q&A]
Dear All,

I am new to NGXS. Online I saw a few examples of how the library is used. My understanding of the general idea is: there is a shared model (as in model-view architecture) and it is capable to notify subscribers through selectors / Observable mechanism about changes in it. Simply integrates with Angular templates. In order to make changes in the state, we dispatch an "action" and handle the update by patchState, for example.

This is all good stuff. What I can't wrap my head around is many times the action of state class will contain logic, for example, to talk to web APIs. Something like below:

```
@Action...
getUser(...)
{
  httpService.get() // Or say, novelService.getNovels(...)
    .subscribe((data) => ctx.patchState());
}
```

Now, I have a problem with this because in my view, state becomes filled with all kinds of business logic and turns into "a swiss knife" of functionality: it is responsible both for model maintenance and all kinds of business rules. I would imagine, the separation of concern principle should dictate that that a dedicated service for business logic should be introduced between state and the consumer of operation.

Another seeming inconvenience of placing business logic inside the state is that state establishes subscribes to observables returned from actions and it is very hard to cancel, for example, http requests that became no longer needed because issuing component is no longer subscribing to http request-sourced observable. I am talking about pipe operator that could not be used from component like this:

`webService.getNovels().pipe(takeUntil(this.ngDestroy$)).subscribe(...)`

So, here goes the thoughts behind my question. Are the examples where business logic is placed inside of state the mainstream approach and if so, why a business logic encapsulation service would not be advisable?

To clarify, pattern I would have in mind mind of how to use state would be like this:

```
onSave(): {
  this.novelService
    .getNovels()  
    .pipe(..)
    .subscribe(novels => this.store.dispatch(new setNovelsAction(novels);
}

```

Thank you very much for reading all this.

**Comment by pdemilly** [comment]
In your @action you don't need to subscribe to your observable. Just return
your observable and let ngxs handle the subscription. If you need to cancel
your subscription check @Action argument  cancelUncompleted

State is a service so you can either do all your business logic in your
actions,  or you can inject another service into the constructor of the
state and call that service from the state

The beauty of ngxs is that your component controllers are now anemic and
only have to deal with minor display functions. 2) you don't have to deal
with subscription/unsubscription anymore. 3) state is immutable 4) no more
cross injection hell in your modules 5) simple bus like architecture 6)
refactoring is a breeze

**Comment by vgpro54321** [comment]
Hello, pdemilly, 

Thank you for taking time to respond.

_In your @action you don't need to subscribe to your observable. Just return
your observable and let ngxs handle the subscription. If you need to cancel
your subscription check @action argument  cancelUncompleted_

Let's compare, for example, http request handling.

1 a. Subscription portion:

No state, plain vanilla http observable
Explicit subscription is required for observable via .subscribe(x => ...)

State action returns observable
Implicit subscription is required for observable via pipe(tap(x => ...))

So, work for subscription portion is pretty identical. As a matter of fact, tap kind takes two function calls to write.

1  b Unsubscribe portion

No state, plain vanilla http observable
Unsubscription is achieved reactively via switchMap,  takeUntil, take etc.  Can action be canceled elegantly in switchMap? Chaining observables with logic inside of state is hard, and those chain elements are often component specific, such as its destruction observable. 

State action returns observable
taken care by framework.

Work is taken care by framework, It seem to be reactive-unfriendly because framework returns hot observable that client side does not have much influence on. This is my main problem and the argument seems to have pro and con sides at least. Framework itself offers un-subscription functionality calling it "action cancellation", so access to the influence on subscription is valuable.

**What are patterns to achieve reactive behaviors with business logic in state?**

_State is a service so you can either do all your business logic in your
actions,  or you can inject another service into the constructor of the
state and call that service from the state_

2  If I implement business logic in state actions, state turns into service that is harder to invoke. Instead of calling method directly, having stack and simplicity programming, I do extra work of declaring action.

If I can inject service into state, I might as well inject it into component directly. Indirection does not produce value.

So, this does not sound as an argument for using business logic in state. 

_1 The beauty of ngxs is that your component controllers are now anemic and
only have to deal with minor display functions. 2) you don't have to deal
with subscription/unsubscription anymore. 3) state is immutable 4) no more
cross injection hell in your modules 5) simple bus like architecture 6)
refactoring is a breeze_

3 
3.1 I completely agree to point 1 that components are simple when they use state as a source of data. I like it as data model.
3.2 I get this as mixed argument: observables are great for reactive programming but framework castrates reactive functionality. I think state consumers could benefit from cold observables.
3.3 Yes, it is, this is good.
3.4 Could you elaborate on what cross injection hell is?
3.5 I don't now if I like "bus" architecture approach in absolute, it is more difficult to debug on another hand because of stack loss, no? Especially when actions trigger other actions and there we have a pack of racing actions.
